### PR TITLE
Use multiple threads for mbus invoke

### PR DESCRIPTION
**Pull request by baldersheim**
@havardpe and @vekterli PR
Only the changes in RPCSend::invoke and RPCnetwork have any semantics. The rest is just due to moving makeLambdaTask.

### CODE REVIEW

File: messagebus/src/vespa/messagebus/network/rpcsend.cpp [hunk 1]
```diff
@@ -236,6 +237,12 @@ void
 RPCSend::invoke(FRT_RPCRequest *req)
 {
     req->Detach();
+    _net->getExecutor().execute(vespalib::makeLambdaTask([&, req]() { doRequest(req);}));
```
Comment (havardpe) [inline messagebus/src/vespa/messagebus/network/rpcsend.cpp]: need to handle rejected tasks

File: staging_vespalib/src/vespa/vespalib/util/lambdatask.h [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 #include <vespa/vespalib/util/executor.h>
 
-namespace search {
+namespace vespalib {
```
Comment (havardpe) [inline staging_vespalib/src/vespa/vespalib/util/lambdatask.h]: please add a unit test for LambdaTask

File: messagebus/src/vespa/messagebus/network/rpcsend.cpp [hunk 3]
```diff
@@ -147,6 +147,11 @@ RPCSend::send(RoutingNode &recipient, const vespalib::Version &version,
 void
 RPCSend::RequestDone(FRT_RPCRequest *req)
 {
+    _net->getExecutor().execute(vespalib::makeLambdaTask([this, req]() { doRequestDone(req);}));
```
Comment (havardpe) [inline messagebus/src/vespa/messagebus/network/rpcsend.cpp]: need to handle rejected tasks

File: messagebus/src/vespa/messagebus/network/rpcnetwork.cpp [hunk 4]
```diff
@@ -382,10 +382,11 @@ RPCNetwork::sync()
 void
 RPCNetwork::shutdown()
 {
-    _executor->shutdown();
     _executor->sync();
     _transport->ShutDown(false);
     _threadPool->Close();
+    _executor->shutdown();
```
Comment (havardpe) [inline messagebus/src/vespa/messagebus/network/rpcnetwork.cpp]: if moving executor shutdown after network shutdown is needed, then the first executor sync is probably just extra waiting...

**Review by havardpe — COMMENTED** [review]
please consider my comments

File: messagebus/src/vespa/messagebus/network/rpcnetwork.cpp (hunk 4 above)
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcnetwork.cpp]: I will keep it there so we get rid of outgoing stuff before closing down network.

File: staging_vespalib/src/vespa/vespalib/util/lambdatask.h (hunk 2 above)
Comment (baldersheim) [reply staging_vespalib/src/vespa/vespalib/util/lambdatask.h]: Done

File: messagebus/src/vespa/messagebus/network/rpcsend.cpp (hunk 3 above)
Comment (baldersheim) [reply messagebus/src/vespa/messagebus/network/rpcsend.cpp]: Asserting that they are not rejected.

**Comment by baldersheim** [comment]
Comments have been addressed, and an undefined test has been fixed.
Will merge once green build

File: messagebus/src/vespa/messagebus/network/rpcsend.cpp (hunk 3 above)
Comment (havardpe) [reply messagebus/src/vespa/messagebus/network/rpcsend.cpp]: The assert was added to the other place where tasks might be rejected

**Review by vekterli — COMMENTED** [review]
:+1:

**Comment by vekterli** [comment]
A bit of a premature ACK on my part, unfortunately. After observing internal system test failures I've remembered at least some of the reasons why we haven't done this earlier.

Long story short, the content layer has several (not explicitly stated) dependencies on the ordering of both requests and replies. These dependencies are only within buckets (or bucket subtrees for splits and joins), but we don't have any explicit bucket message sequencing channels today. MessageBus historically guarantees request and reply ordering when only a single thread is used.

Likely dependencies that I'm aware of:
* Applying BucketInfo replies to the distributor bucket database assumes that the apply order for a particular node index is equal to the order in which the operations were actually executed on the content node. Failure to do so may cause bucket databases to get out of sync with the actual state of the content nodes.
* Content nodes do a read-modify-write of the max observed mutation timestamp for all incoming mutations to a bucket. If an "outdated" timestamp is observed, the message is rejected. If mutations sent from a distributor are reordered before this point, the timestamp check may fail arbitrarily.

**Pull request merged**